source/common/filesystem/subscription_impl.h

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include "envoy/config/subscription.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/filesystem/filesystem.h"
+
+#include "common/common/macros.h"
+#include "common/config/utility.h"
+#include "common/filesystem/filesystem_impl.h"
+
+#include "api/base.pb.h"
+#include "google/protobuf/util/json_util.h"
+
+namespace Envoy {
+namespace Filesystem {
+
+/**
+ * Filesystem inotify implementation of the API Subscription interface. This allows the API to be
+ * consumed on filesystem changes to files containing the JSON canonical representation of
+ * lists of ResourceType.
+ */
+template <class ResourceType> class SubscriptionImpl : public Config::Subscription<ResourceType> {
+public:
+  SubscriptionImpl(Event::Dispatcher& dispatcher, const std::string& path)
+      : path_(path), watcher_(dispatcher.createFilesystemWatcher()) {}
+
+  // Config::Subscription
+  void start(const std::vector<std::string>& resources,
+             Config::SubscriptionCallbacks<ResourceType>& callbacks) override {
+    // We report all discovered resources in the watched file.
+    UNREFERENCED_PARAMETER(resources);
+    callbacks_ = &callbacks;
```

Comment:
Got it (I suppose you could solve this with Optional too, but that can get ugly). It might be good to add a comment on the member that it's a non-owned pointer.